lib/tbot/loop.go

```diff
@@ -0,0 +1,146 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package tbot
+
+import (
+	"context"
+	"log/slog"
+	"math"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+)
+
+type runOnIntervalConfig struct {
+	name                 string
+	f                    func(ctx context.Context) error
+	clock                clockwork.Clock
+	reloadCh             chan struct{}
+	log                  *slog.Logger
+	interval             time.Duration
+	retryLimit           int
+	exitOnRetryExhausted bool
+	waitBeforeFirstRun   bool
+}
+
+// runOnInterval runs a function on a given interval, with retries and jitter.
+//
+// TODO(noah): Emit Prometheus metrics for:
+// - Success/Failure of attempts
+// - Time taken to execute attempt
+// - Time of next attempt
+func runOnInterval(ctx context.Context, cfg runOnIntervalConfig) error {
+	switch {
+	case cfg.interval <= 0:
+		return trace.BadParameter("interval must be greater than 0")
+	case cfg.retryLimit < 0:
+		return trace.BadParameter("retryLimit must be greater than or equal to 0")
+	case cfg.log == nil:
+		return trace.BadParameter("log is required")
+	case cfg.f == nil:
+		return trace.BadParameter("f is required")
+	case cfg.name == "":
+		return trace.BadParameter("name is required")
+	}
+
+	// If no reload channel is provided, create one that never yields a value.
+	if cfg.reloadCh == nil {
+		cfg.reloadCh = make(chan struct{})
+	}
+	log := cfg.log.With("task", cfg.name)
+
+	if cfg.clock == nil {
+		cfg.clock = clockwork.NewRealClock()
+	}
+
+	ticker := cfg.clock.NewTicker(cfg.interval)
+	defer ticker.Stop()
+	jitter := retryutils.NewJitter()
+	firstRun := true
+	for {
+		if !firstRun || (firstRun && cfg.waitBeforeFirstRun) {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-ticker.Chan():
+			case <-cfg.reloadCh:
+			}
+		}
+		firstRun = false
+
+		var err error
+		for attempt := 1; attempt <= cfg.retryLimit; attempt++ {
+			log.InfoContext(
+				ctx,
+				"Attempting task",
+				"attempt", attempt,
+				"retry_limit", cfg.retryLimit,
+			)
+			err = cfg.f(ctx)
+			if err == nil {
+				break
+			}
+
+			if attempt != cfg.retryLimit {
+				// exponentially back off with jitter, starting at 1 second.
+				backoffTime := time.Second * time.Duration(math.Pow(2, float64(attempt-1)))
```

Comment:
This is fine, but I wonder if there's any use case that might benefit from a configurable initial duration other than `time.Second`?

Reply:
I went with this for now to just keep the existing behaviour - I figure once we know what we want to do to customise this (e.g more this more configurable w/ the option for a linear instead of expo), we can go ahead and modify it then,.